Add S3 to HDFS File copy functionality.

caskchester opened the pull request:
1. Adds a file metadata source template that can be extended to read
   file metadata from any source filesystem.

2. Adds a file copy sink template that can be exended to copy files
   from a remote source given file metadata as inputs.

3. Added InputFormat and OutputFormat to support source and sink plugins.

Commit:
Add S3 to HDFS File copy functionality.

1. Adds a file metadata source template that can be extended to read
   file metadata from any source filesystem.

2. Adds a file copy sink template that can be exended to copy files
   from a remote source given file metadata as inputs.

3. Added InputFormat and OutputFormat to support source and sink plugins.

caskchester commented:
Design doc: https://wiki.cask.co/pages/viewpage.action?pageId=4364388

albertshau commented on core-plugins/docs/HDFSFileCopySink-batchsink.md:
```diff
@@ -0,0 +1,37 @@
+| **Reference Name**     |  **Y**   | None      | This will be used to uniquely identify this source for lineage, annotating metadata, etc.   |
+| **Base Path**          |  **Y**   | None      | The folder where the copied files will be placed. It will be created if it doesn't exist.   |
+| **Enable Overwrite**   |  **Y**   | False     | Specifies whether or not to overwrite files if it already exists.                           |
+| **Buffer Size**        |  **Y**   | True      | The size of the buffer that temporarily stores data from file input stream while copying.   |
```
default should be a number right?

albertshau commented on core-plugins/docs/HDFSFileCopySink-batchsink.md:
```diff
@@ -0,0 +1,37 @@
+
+Usage Notes
+-----------
+This sink plugin only reads StructuredRecords with the follwing schema. Each record should contain the metadata for a file to be copied
```
typo: 'following'

albertshau commented on core-plugins/docs/HDFSFileCopySink-batchsink.md:
```diff
@@ -0,0 +1,37 @@
+| :--------------------- | :----- | :-------------------------                                                                                                                     |
+| **File Name**          | String | Only contains the name of the file.                                                                                                            |
+| **Full Path**          | String | Contains the full path of the file in the source file system.                                                                                  |
+| **File Size**          | long   | Specifies the number of fi                                                                                                                     |
```
incomplete description

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileCopySink.java:
```diff
@@ -0,0 +1,144 @@
+    @Macro
+    @Nullable
+    @Description("The size of the buffer that temporarily stores data from file input stream.")
+    public int bufferSize;
```
in order for this to be Nullable, it should be an Integer instead of an int.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileMetadata.java:
```diff
@@ -0,0 +1,170 @@
+  public static final String PERMISSION = "permission";
+
+  // contains only the name of the file
+  protected String fileName;
```
when possible, these should be final. It's easier to maintain and use objects if they are immutable.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileMetadata.java:
```diff
@@ -0,0 +1,170 @@
+  protected String owner;
+
+  // whether or not the file is a folder
+  protected Boolean isFolder;
```
this can be a boolean instead of Boolean so there is no null possibility.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileMetadata.java:
```diff
@@ -0,0 +1,170 @@
+  public abstract Credentials getCredentials();
+
+  public Boolean getIsFolder() {
+    return this.isFolder;
```
nit: for consistency, `return isFolder `

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileMetadata.java:
```diff
@@ -0,0 +1,170 @@
+
+  public abstract Credentials getCredentials();
+
+  public Boolean getIsFolder() {
```
nit: can name this isFolder()

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileMetadata.java:
```diff
@@ -0,0 +1,170 @@
+  public AbstractFileMetadata(FileStatus fileStatus, String sourcePath, Credentials credentials) {
+    String fullPath = fileStatus.getPath().toString();
+    String[] paths = fullPath.split("/");
+    this.fileName = paths[paths.length - 1];
```
the Path object from fileStatus.getPath() has a getName(). No need to split on the full path.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileMetadata.java:
```diff
@@ -0,0 +1,170 @@
+  protected long fileSize;
+
+  // modification time of file
+  protected long timeStamp;
```
nit: should probably name this modificationTime or something more descriptive.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileCopySink.java:
```diff
@@ -0,0 +1,144 @@
+    AbstractFileMetadata output;
+    switch ((String) input.get(AbstractFileMetadata.Credentials.DATA_BASE_TYPE)) {
+      case S3FileMetadata.DATA_BASE_NAME :
+        output = new S3FileMetadata((String) input.get(AbstractFileMetadata.FILE_NAME),
```
I think it's better to put the AbstractFileMetadata to and from StructuredRecord conversion logic inside the AbstractFileMetadata class, through a `toRecord()` method and a `static fromRecord(StructuredRecord record)` method.
Then the logic is in a single place, it can be re-used in other classes, and the field names can be private

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileMetadata.java:
```diff
@@ -0,0 +1,170 @@
+   * extend from this class to create a credential class for each filesystem
+   */
+  public abstract static class Credentials {
+    public String databaseType;
```
this isn't really a credential, it seems like another attribute of the file. 'database' is also a loaded term, as people usually think of sql based relational databases. We should use 'fileSystem' instead.

caskchester replied:
The information required to access the source filesystem will be stored in its subclass, so this class is called credentials.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileMetadata.java:
```diff
@@ -0,0 +1,170 @@
+
+    // TODO: investigate how to cleanly set basePath
+    if (sourcePath.equals("")) {
+      basePath = this.fileName;
```
what does this mean? I'm confused about what basePath is. It's not the directory the files came from?

caskchester replied:
Example: suppose there is a file in `/a/b/c/d/haha.txt`
If you do `cp  -r   /a/b/c   /dest/`
Base path in this case would be `c/d/haha.txt`. Therefore the resulting file path would be `/dest/c/d/haha.txt`

I'll add a comment in the code to explain it more clearly.

albertshau replied:
I think relativePath would be a better name for this. I would expect a base path to be the left part not the right part. It would also help to have a comment.

albertshau replied:
The logic seems like it can be simpler. I believe you can do `fileStatus.getPath().toURI().getPath()` to get the path without scheme, host, port. If I understand correctly, that path should always be prefixed by sourcePath? If so, you can do `fileStatus.getPath().toURI().getPath().substring(sourcePath.length())` (might be off by one).

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileMetadata.java:
```diff
@@ -0,0 +1,170 @@
+       * this block of code drops the URI and reconstructs basePath by
+       * concatenating the folders with "/"
+       */
+      String pathWithoutURI = fullPath.substring(fullPath.indexOf(sourcePath));
```
is this trying to strip the scheme (change hdfs:///abc to /abc)? you can use the Path object's toURI().getPath()

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileMetadata.java:
```diff
@@ -0,0 +1,170 @@
+      for (int i = numStrip; i < pathsWithoutURI.length; i++) {
+        basePath = basePath.concat(pathsWithoutURI[i] + "/");
+      }
+      if (basePath.length() > 0) {
```
and basePath is the sourcePath with the front stripped as well?

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileMetadata.java:
```diff
@@ -0,0 +1,170 @@
+    }
+  }
+
+  public AbstractFileMetadata(String fileName, String fullPath, long timeStamp, String owner,
```
abstract classes should have protected constructors.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileMetadata.java:
```diff
@@ -0,0 +1,170 @@
+  }
+
+  public AbstractFileMetadata(String fileName, String fullPath, long timeStamp, String owner,
+                              Long fileSize, Boolean isFolder, String basePath,
```
Long and Boolean should be long and boolean

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileMetadataSource.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    public void validate() {
+      }
```
fix indentation

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileMetadataSource.java:
```diff
@@ -0,0 +1,114 @@
+
+    @Macro
+    @Description("The number of files each split manipulates")
+    public int maxSplitSize;
```
Macro enabled properties should be Integer, as it will be empty at configure time when there is nothing to substitute.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputFormat.java:
```diff
@@ -0,0 +1,171 @@
+
+  protected static final String SOURCE_PATHS = "source.paths";
+  protected static final String MAX_SPLIT_SIZE = "max.split.size";
+  protected static final String FS_URI = "uri";
```
filesystem.uri would be more clear

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputFormat.java:
```diff
@@ -0,0 +1,171 @@
+    Configuration conf = jobContext.getConfiguration();
+    URI uri = URI.create(conf.get(FS_URI));
+    String[] sourcePaths = conf.get(SOURCE_PATHS).split(",");
+    Boolean recursiveCopy = conf.getBoolean(RECURSIVE_COPY, true);
```
boolean, since it can't be null

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputFormat.java:
```diff
@@ -0,0 +1,171 @@
+    return recordReader;
+  }
+
+  private void recursivelyAddFileStatus (List<FileStatus> fileStatuses, List<String> baseFolders,
```
remove extra space after AddFileStatus

albertshau replied:
FileSystem also has a listFiles(Path, boolean) method that can list recursively. Should use that instead of your own unless there is some missing functionality with it.

caskchester replied:
`listFiles(Path, boolean)` ignores empty folders, so I wrote my own method.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputFormat.java:
```diff
@@ -0,0 +1,171 @@
+          prefix = "";
+        }
+        recursivelyAddFileStatus(fileStatuses, baseFolders, prefix, fileStatus, recursiveCopy, fileSystem);
+      } catch (Exception e) {
```
exceptions need to be handled better. For example, you should probably let a FileNotFoundException on the source path bubble up and fail the job. In general, it seems like the IOExceptions that can be thrown here should cause the job to fail, not logged and skipped.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputFormat.java:
```diff
@@ -0,0 +1,171 @@
+                                                                        getCredentialsFromConf(conf));
+
+    // compute number of splits and instantiate the splits
+    int numSplits = (fileMetaDataList.size() - 1) / maxSplitSize + 1;
```
how come it's subtracting and adding? if there are 10 files and maxSplitSize is 2, shouldn't there be 5 splits and not 3?

caskchester replied:
Yeah if you plug those numbers in you will get 5.

albertshau replied:
right, i seem to have inserted parentheses in my head

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputFormat.java:
```diff
@@ -0,0 +1,171 @@
+    for (int i = 0; i < fileMetaDataList.size(); i++) {
+      AbstractMetadataInputSplit minInputSplit = abstractInputSplits.poll();
+      minInputSplit.addFileMetadata(fileMetaDataList.get(i));
+      abstractInputSplits.add(minInputSplit);
```
this doesn't enforce maxSplitSize. Should check the number of files in the split before adding it back to the PriorityQueue. If it's greater than max, add it to inputSplits (from below) and not back to the PriorityQueue.

caskchester replied:
Yeah that's one of the concerns I had when I wrote this method. Consider mapping the following list of files to only 2 splits.
one 100GB file ....... 1000 50B files ......... one 100GB file.
If we enforce the number of files in maxSplitSize, the two big files would end up in the same split.

I agree that the pipeline should behave how the users expect it to behave. Perhaps sorting the files by size first would be a better idea?

albertshau replied:
True, you can sort those first and it would probably be more evenly distributed.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputFormat.java:
```diff
@@ -0,0 +1,171 @@
+    }
+  }
+
+  private List<AbstractFileMetadata> getFileMetadaDataList (List<FileStatus> fileStatuses,
```
typo: Metada

Also, remove extra space after DataList

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputFormat.java:
```diff
@@ -0,0 +1,171 @@
+    fileSystem.close();
+
+    // create a filemetadata list of all the files
+    List<AbstractFileMetadata> fileMetaDataList = getFileMetadaDataList(fileStatuses, baseFolders,
```
seems like you should be able to build this list while getting each FileStatus instead of making another pass. Would also make the logic simpler, as you wouldn't have to maintain an extra baseFolders list.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputFormat.java:
```diff
@@ -0,0 +1,171 @@
+                                                          String sourcePath,
+                                                          AbstractFileMetadata.Credentials credentials);
+
+  public static void setSourcePaths(Job job, String value) {
```
nit: can make it take a Configuration object instead.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputSplit.java:
```diff
@@ -0,0 +1,162 @@
+    AbstractMetadataInputSplit other = (AbstractMetadataInputSplit) o;
+    long myLength = getTotalSize();
+    long otherLength = other.getTotalSize();
+    if (myLength < otherLength) {
```
slightly easier to do `Long.compare(x, y)`

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputSplit.java:
```diff
@@ -0,0 +1,162 @@
+        writeCredentials(dataOutput, fileMetaData.getCredentials());
+      }
+
+    } catch (Exception e) {
```
shouldn't log and skip. Just let the exception propagate, it should be handled by mapreduce.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputSplit.java:
```diff
@@ -0,0 +1,162 @@
+    try {
+      // write obj summaries
+      dataOutput.writeLong(this.getLength());
+      for (int i = 0; i < this.getLength(); i++) {
```
In general, to iterate through collections you want to do

```
for (AbstractFileMetadata fileMetaData : fileMetaDataList) {
```

For example, if it's not an ArrayList, `.get(i)` could be more expensive than iterating through items.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputSplit.java:
```diff
@@ -0,0 +1,162 @@
+        // convert each filestatus (serializable) to byte array
+        AbstractFileMetadata fileMetaData = fileMetaDataList.get(i);
+
+        dataOutput.writeUTF(fileMetaData.getFileName());
```
seems like AbstractFileMetadata should also be Writable, and you just call its write(dataOutput) method here. Then the logic will be contained in AbstractFileMetadata. Otherwise, changing AbstractFileMetadata would also require changing this class.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/FileCopyRecordWriter.java:
```diff
@@ -0,0 +1,130 @@
+  }
+
+  // source filesystems
+  private Map<AbstractFileMetadata.Credentials, FileSystem> sourceFilesystemMap;
```
nit: put this before the constructor.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/s3/S3FileMetadataSource.java:
```diff
@@ -0,0 +1,153 @@
+  );
+
+  private Schema outputSchema;
+
```
there should be some validation done on the config in the configurePipeline method. This ensures that invalid configs cause the pipeline to fail to deploy instead of waiting until runtime to fail.

albertshau replied:
validation should be basic things like checking max size is not negative, checking the path starts with a valid scheme, etc.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/FileCopyRecordWriter.java:
```diff
@@ -0,0 +1,130 @@
+  @Override
+  public void write(NullWritable key, AbstractFileMetadata fileMetadata) throws IOException, InterruptedException {
+    // get source database connection
+    AbstractFileMetadata.Credentials sourceCredentials = fileMetadata.getCredentials();
```
it seems insecure to have the credentials in every record. We can address this in a follow up PR, but maybe the sink should require the source credentials in its configuration as well.

caskchester replied:
This is a design choice. Ideally a file copy sink can take inputs from multiple sources (different filesystems). That's why we attach the credentials to every file metadata record we send. But yeah I agree it could be a security concern.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/FileCopyRecordWriter.java:
```diff
@@ -0,0 +1,130 @@
+
+    // construct file paths for source and destination
+    Path srcPath = new Path(fileMetadata.fullPath);
+    Path destPath = new Path(basePath + "/" + fileMetadata.basePath);
```
shouldn't manually concatenate, use `new Path(basePath, fileMetadata.basePath)`

albertshau replied:
the terminology is also confusing. So fileMetadata.basePath is the directory relative to the path given to the source? And basePath here is the destination directory?

albertshau replied:
can address this in a follow up PR, but data should be written to a temporary directory that the OutputCommitter renames at the very end. This is because MapReduce tries to be atomic if possible.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/FileCopyRecordWriter.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    // immediately return if we dont want to overwrite and file exists in destination
+    if (!enableOverwrite && destFileSystem.exists(destPath)) {
```
you have to be a little careful here. If the mapper died early and is reprocessing this file, the file may exist but not be complete. Similarly, if speculative execution is enabled, mapreduce will launch multiple mappers on the same input split, and they should not conflict with each other. Basically, each mapper should write to its own unique directory in order to handle these scenarios.

This can be addressed in a follow up PR.

caskchester replied:
Doesn't [FileOutputCommitter](https://hadoop.apache.org/docs/r2.7.2/api/org/apache/hadoop/mapred/FileOutputCommitter.html) prevent failed jobs from being committed?

albertshau replied:
I think it does some setup and cleanup, but you may need to make sure you're using whatever directories it is setting/cleaning up instead of the destination one specified by the plugin.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/FileCopyRecordWriter.java:
```diff
@@ -0,0 +1,130 @@
+    while ((len = inputStream.read(buf)) >= 0) {
+      outputStream.write(buf, 0, len);
+    }
+    inputStream.close();
```
in general, close should always be done in a finally block, in case an exception is thrown earlier.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/FileCopyRecordWriter.java:
```diff
@@ -0,0 +1,130 @@
+  @Override
+  public void close(TaskAttemptContext taskAttemptContext) throws IOException, InterruptedException {
+    /*
+     * TODO: investigate when to close the destination filesystem and the source filesystems
```
this is closing the record writer, so it should be safe to close everything here. The record writer won't be used anymore after this.

caskchester replied:
When I close connections in the `close()` method, a pipeline with multiple splits would fail. It seems like some splits are sharing one `RecordWriter` instantiation. So if one split calls this method to close the connection, another split that's also using this `RecordWriter` would fail during its call to the `write()` method.

The connections are stored in a `Map<Credentials, Filesystem>` object

albertshau replied:
hm I think that means the FileSystem you're getting is a cached instance, not that the RecordWriter is shared between splits. If you look at FileSystem.get(), it keeps an internal cache as well, which can be turned off depending on the value of a config value:

```
    String disableCacheName = String.format("fs.%s.impl.disable.cache", scheme);
    if (conf.getBoolean(disableCacheName, false)) {
      return createFileSystem(uri, conf);
    }
```

caskchester replied:
I see. I guess this means we have some questions to figure out:
1. If we turn caching off, is it okay to have 100 splits all try to initiate connection to the filesystem at the same time?
2. If we leave caching on, does that mean file transmission resources (like bandwidth and other things) would be divided among the splits?

Maybe we can address these problems in a future PR.

albertshau replied:
Were you seeing the issue in a cluster or just in the sandbox/unit test? I wouldn't expect to see this in a cluster where the code is run in separate processes.

caskchester replied:
Yeah the issue only happens in sandbox if caching is not disabled. Doesn't happen in cluster

albertshau reviewed (commented):
looks pretty good, have some comments plus some points we should address in future work

Commit:
Addressed comments.

1. Made all members in AbstractFileMetadata private and final.
2. Removed AbstractFileMetadata.Credentials class. Additional credentials
   directly stored in classes extended from AbstractFileMetadata.
3. Added non-editable Schema in UI.
4. Added two options in FileCopySink: use cached filesystem and preserve file owner
5. Replaced FileOutputCommitter with empty committer.
6. Reordered execution in FileCopyRecordWriter for better copying performance.

caskchester commented:
@albertshau @sagarkapare 
Addressed comments. Please take another look. Thanks!

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileCopySink.java:
```diff
@@ -85,32 +80,60 @@ public void transform(StructuredRecord input,
+    // TODO: figure out why this still shows up as required in configuration UI
     @Macro
     @Nullable
     @Description("The size of the buffer that temporarily stores data from file input stream.")
```
the description should mention what unit this is in. MB? If it's currently in bytes, we should probably change it to megabytes, as it's pretty annoying putting in byte values.

albertshau replied:
Description should also mention what it defaults to.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileCopySink.java:
```diff
@@ -85,32 +80,60 @@ public void transform(StructuredRecord input,
     public Boolean enableOverwrite;
 
+    @Description("Whether or not for splits to use cached connections to the filesystems.")
+    public Boolean filesystemCaching;
```
I don't think we should have this property. It's misleading, a FileSystem doesn't represent a connection to the filesystem.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileMetadata.java:
```diff
@@ -160,11 +233,78 @@ public short getPermission() {
+
   /**
-   * extend from this class to create a credential class for each filesystem
+   * Don't use this to deserialize AbstractFileMetadata. Use the constructor
```
implementing this is part of the contract for the Writable interface. Hadoop calls this method to deserialize the class when it is used as the mapper output key or value, or when SequenceFileOutputFormat is used. If this isn't implemented, it won't work correctly when used in that way.

Since we're not using this as a mapper output key or value, I think it's fine to remove WritableComparable as a parent interface so you aren't forced to implement this unused method.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputFormat.java:
```diff
@@ -118,54 +122,37 @@ public RecordReader createRecordReader(InputSplit inputSplit, TaskAttemptContext
   }
 
-  private void recursivelyAddFileStatus (List<FileStatus> fileStatuses, List<String> baseFolders,
+  private void recursivelyAddFileStatus(List<AbstractFileMetadata> fileMetadataList,
```
lets add a comment saying this is used instead of FileSystem.listFiles() because that method doesn't list empty directories.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputFormat.java:
```diff
@@ -118,54 +122,37 @@ public RecordReader createRecordReader(InputSplit inputSplit, TaskAttemptContext
 
-  private void recursivelyAddFileStatus (List<FileStatus> fileStatuses, List<String> baseFolders,
+  private void recursivelyAddFileStatus(List<AbstractFileMetadata> fileMetadataList,
                                          String prefix, FileStatus fileStatus, Boolean enable,
```
nit: rename enable to 'recursive'

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputSplit.java:
```diff
@@ -52,56 +52,24 @@ public void write(DataOutput dataOutput) throws IOException {
 
-    } catch (Exception e) {
-      LOG.error("serialization failed");
+    } catch (IOException e) {
```
don't need this catch clause

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/FileCopyRecordWriter.java:
```diff
@@ -41,88 +41,128 @@
-    AbstractFileMetadata.Credentials sourceCredentials = fileMetadata.getCredentials();
-    if (!sourceFilesystemMap.containsKey(sourceCredentials)) {
-      sourceFilesystemMap.put(sourceCredentials, getSourceFilesystemFromCredentials(sourceCredentials));
+  public FileCopyRecordWriter(Configuration conf) throws Exception {
```
should throw as specific an exception as possible. I think this should be IOException.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/FileCopyOutputFormat.java:
```diff
@@ -74,7 +112,7 @@ public RecordWriter getRecordWriter(TaskAttemptContext taskAttemptContext) throw
     try {
       return new FileCopyRecordWriter(conf);
     } catch (Exception e) {
```
don't need this try catch if FileCopyRecordWriter throws IOException

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/FileCopyRecordWriter.java:
```diff
@@ -41,88 +41,128 @@
+    if (fileMetadata.getBasePath().isEmpty()) {
+      // nothing to create
+      return;
+    } else {
```
don't need to put this in an else clause, can just declare and assign it here.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/FileCopyRecordWriter.java:
```diff
@@ -41,88 +41,128 @@
+      if (preserveOwner) {
+        destFileSystem.setOwner(destPath, fileMetadata.getOwner(), fileMetadata.getGroup());
+      }
+      inputStream.close();
```
if this throws an exception, outputStream won't get closed. So you need to wrap this in a try catch finally as well.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/FileCopyRecordWriter.java:
```diff
@@ -41,88 +41,128 @@
+    URI uri = URI.create(metadata.getHostURI());
+
+    // whether or not to disable caching for source filesystems
+    String disableCacheName = String.format("fs.%s.impl.disable.cache", uri.getScheme());
```
lets just always disable cache for now.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/s3/S3FileMetadataSource.java:
```diff
@@ -137,17 +101,22 @@ public void transform(KeyValue<NullWritable, S3FileMetadata> input, Emitter<Stru
+    public void validate() {
+      super.validate();
+      if (!this.containsMacro("filesystemURI")) {
+        if (!filesystemURI.startsWith("s3a://")) {
```
s3n is an older, but still supported scheme. should allow it too.

caskchester replied:
Can I just replace all `s3n` URIs with `s3a`? This is from the [AWS doc:](https://aws.amazon.com/premiumsupport/knowledge-center/emr-file-system-s3/) 

> For Apache Hadoop, S3A is the successor to S3N and is backward compatible with S3N. Using Apache Hadoop, all objects accessible from s3n:// URLs should also be accessible from S3A by replacing the URL scheme.

albertshau replied:
I'm not sure when they replaced it, but we have to support some old Hadoop distributions so you can only do that if we verify that s3a is supported in all of our supported Hadoop distributions.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileMetadata.java:
```diff
@@ -160,11 +233,78 @@ public short getPermission() {
+
+  protected abstract void addCredentialsToBuilder(StructuredRecord.Builder builder);
+
+  protected abstract String getFSName();
```
where/how is this used? Is there a reason this is used instead of the scheme for the filesystem uri? (hdfs, s3a, s3n, ftp, etc.)

caskchester replied:
yeah it's redundant. will remove

caskchester commented:
@albertshau @sagarkapare 
Addressed comments and added support for s3n.
Please review. Thanks!

Commit:
Addressed comments #2
1. Added support for s3n.
2. Improved exception handling.
3. removed redundant field (filesystem) from filemetadata.
4. improved AbstractMetadataInputSplit.recursivelyAddFileStatus method.
5. use one line of code to set relative path in abstractfilemetadata

caskchester commented:
@albertshau @sagarkapare
Addressed comments and added support for s3n.
Please review. Thanks!

albertshau commented on core-plugins/docs/HDFSFileCopySink-batchsink.md:
```diff
@@ -36,7 +35,7 @@ This sink plugin only reads StructuredRecords with the following schema. Each re
 | **owner**              | String | The owner of the file.                                                                                                                         |
 | **isFolder**           | Boolean| Whether or not the file is a folder.                                                                                                           |
-| **basePath**           | String | The base path of the file. This path will be appended to the base path specified in the FileCopySink to create the file in the destination.    |
+| **relativePath**       | String | The relavite path is constructed by deleting the portion of the source path that comes before the last path separator ("/") from the full path.|
```
typo, relavite -> relative

albertshau commented on core-plugins/docs/S3FileMetadataSource-batchsource.md:
```diff
@@ -39,7 +39,7 @@ A structured record with the following schema is emitted for each file it reads.
 | **owner**              | String | The owner of the file.                                                                                                                         |
 | **isFolder**           | Boolean| Whether or not the file is a folder.                                                                                                           |
-| **basePath**           | String | The base path of the file. This path will be appended to the base path specified in the FileCopySink to create the file in the destination.    |
+| **relativePath**       | String | The relavite path is constructed by deleting the portion of the source path that comes before the last path separator ("/") from the full path.|
```
typo, relavite

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileCopySink.java:
```diff
@@ -147,8 +147,16 @@ public FileCopyOutputFormatProvider(AbstractFileCopySink.AbstractFileCopySinkCon
+      }
+
+      // always disable caching
+      conf.put(String.format("fs.%s.impl.disable.cache", config.getScheme()), String.valueOf(true));
```
We only want our specific OutputFormat to disable caching, not everything else in Hadoop. So we shouldn't set this here, but only in the OutputFormat when we get a new FileSystem object.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputSplit.java:
```diff
@@ -50,41 +56,53 @@ public AbstractMetadataInputSplit() {
       dataOutput.writeLong(this.getLength());
+
+      // write total number of file bytes
+      dataOutput.writeLong(this.getTotalBytes());
```
this works, but most of the time you write the length of a collection and immediately follow it with the items in the collection. Let's do that here too and write the total bytes before we write the number of files.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputSplit.java:
```diff
@@ -50,41 +56,53 @@ public AbstractMetadataInputSplit() {
     long numObjects = dataInput.readLong();
+
+    // get total number of bytes
+    totalBytes = dataInput.readLong();
```
move this above reading the number of files when write() is changed.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputSplit.java:
```diff
@@ -50,41 +56,53 @@ public AbstractMetadataInputSplit() {
-      fileMetaDataList.add(readFileMetaData(dataInput));
+      AbstractFileMetadata metadata = readFileMetaData(dataInput);
+      fileMetaDataList.add(metadata);
+      totalBytes += metadata.getFileSize();
```
how come you need to add it here? doesn't this mean totalBytes will be doubled?

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/FileCopyRecordWriter.java:
```diff
@@ -121,50 +120,72 @@ public void write(NullWritable key, AbstractFileMetadata fileMetadata) throws IO
+   * @param fs The iterator over all the filesystems we wish to close.
+   * @throws IOException
+   */
+  private void safelyCloseSourceFilesystems(Iterator<Map.Entry<String, FileSystem>> fs) throws IOException {
```
since the map key isn't used, you can take an `Iterator<FileSystem>` instead and pass in `sourceFilesystemMap.values().iterator()`.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/FileCopyRecordWriter.java:
```diff
@@ -121,50 +120,72 @@ public void write(NullWritable key, AbstractFileMetadata fileMetadata) throws IO
+      try {
+        fs.next().getValue().close();
+      } finally {
+        safelyCloseSourceFilesystems(fs);
```
its a little weird to go through an iterator using recursion instead of a while loop. Let's just do
```
while (fs.hasNext()) {
  try {
    fs.next().close();
  } catch () {
    // log some warning
  }
}
```

albertshau commented on core-plugins/widgets/HDFSFileCopySink-batchsink.json:
```diff
@@ -41,20 +41,8 @@
-          "label": "Buffer Size",
+          "label": "Buffer Size (MB)",
           "name": "bufferSize",
           "default": "4096"
```
default it to 1 to be consistent with code

albertshau reviewed (commented):
just a few style comments remaining

Commit:
Addressed comments #3
1. Typo fixes.
2. Filesystem cache disabling now only happens in FileCopyRecordWriter.
3. Logs a warning and skip if it fails to close source Filesystems.

caskchester commented:
@albertshau 
Addressed comments. Please review. Thanks!

albertshau reviewed (approved):
lgtm

albertshau commented:
i think we wanted to add this as a separate repo under the hydrator org, so lets not merge it here

caskchester commented:
Closing this and moving to https://github.com/hydrator/whole-file-ingest per @nitinmotgi 's request
